cdap-watchdog/src/main/java/co/cask/cdap/logging/plugins/RollingLocationLogAppender.java

```diff
@@ -0,0 +1,274 @@
+import co.cask.cdap.proto.id.NamespaceId;
+import com.google.common.annotations.VisibleForTesting;
+import com.google.inject.Inject;
+import org.apache.twill.filesystem.Location;
+import org.apache.twill.filesystem.LocationFactory;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.io.Closeable;
+import java.io.Flushable;
+import java.io.IOException;
+import java.io.OutputStream;
+
+import static ch.qos.logback.core.CoreConstants.CODES_URL;
+
+/**
+ * Rolling Appender for {@link Location}
+ */
+public class RollingLocationLogAppender extends FileAppender<ILoggingEvent> implements Flushable {
+  private static final Logger LOG = LoggerFactory.getLogger(RollingLocationLogAppender.class);
+
+  // TODO CDAP-8196 LocationFactory should be passed from Log framework
+  @Inject
+  private LocationFactory locationFactory;
+
+  private TriggeringPolicy<ILoggingEvent> triggeringPolicy;
+  private RollingPolicy rollingPolicy;
+
+  private String basePath;
+  private String filePath;
+  private String filePermissions;
+  private LocationManager locationManager;
+
+  public RollingLocationLogAppender() {
+    setName(getClass().getName());
+  }
+
+  @Override
+  public void start() {
+    if (validateAppenderProperties()) {
+      return;
+    }
+
+    // TODO remove if condition after removing Guice injection
+    if (locationFactory != null) {
+      try {
+        this.locationManager = new LocationManager(locationFactory, basePath, filePermissions);
+      } catch (IOException e) {
+        // not handling it since we will remove this if condition after log framework is in place
+      }
+      started = true;
+    }
+  }
+
+  @Override
+  public void doAppend(ILoggingEvent eventObject) throws LogbackException {
+    try {
+      // TODO Remove 'if' after removing Guice injection
+      if (locationFactory != null && this.locationManager == null) {
+        this.locationManager = new LocationManager(locationFactory, basePath, filePermissions);
+        started = true;
+      }
+
+      // logic from AppenderBase
+      if (!this.started) {
+        addStatus(new WarnStatus("Attempted to append to non started appender [" + name + "].", this));
+        return;
+      }
+
+      // logic from AppenderBase
+      if (getFilterChainDecision(eventObject) == FilterReply.DENY) {
+        return;
+      }
+
+      String namespaceId = eventObject.getMDCPropertyMap().get(LocationManager.TAG_NAMESPACE_ID);
+      if (!namespaceId.equals(NamespaceId.SYSTEM.getEntityName())) {
+        LocationIdentifier logLocationIdentifier = locationManager.getLocationIdentifier(eventObject
+                                                                                           .getMDCPropertyMap());
+        rollover(logLocationIdentifier, eventObject);
+        OutputStream locationOutputStream = locationManager.getLocationOutputStream(logLocationIdentifier, filePath);
+        setOutputStream(locationOutputStream);
+        writeOut(eventObject);
```

Comment:
so `setOutputStream` return if encoder is not set (null), however writeOut just tries to write using the encoder and that would cause NPE right ?

also we have the output stream opened, even if encoder is not set, though encoder is the only person to write to output stream here.